Get Frontera support working again

wthrowe opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Remove redundant target and set MPI{CC,CXX}

The charm++ target is included in the LIBS target.

The charm build system requires the mpi compilers to be specified when
using mpi.

wthrowe committed:
Use BFD linker on frontera

wthrowe committed:
Update Frontera submit script for queue changes

wthrowe committed:
Move CPATH to {C,CPLUS}_INCLUDE_PATH on Frontera

Frontera part of #3699.

nilsvu commented on support/Environments/frontera_gcc.sh:
```diff
@@ -181,20 +182,23 @@ EOF
-        ./build charm++ mpi-linux-x86_64 smp mpicxx --with-production -j6
-        ./build LIBS mpi-linux-x86_64 smp mpicxx --with-production -j6
+        MPICC=mpicc MPICXX=mpicxx \
+             ./build charm++ mpi-linux-x86_64 smp mpicxx --with-production -j6
```
You can delete this line. The `LIBS` target includes the `charm++` target, see [docs](https://github.com/UIUC-PPL/charm#building-the-source)

nilsvu commented on support/Environments/frontera_gcc.sh:
```diff
@@ -181,20 +182,23 @@ EOF
+        MPICC=mpicc MPICXX=mpicxx \
+             ./build charm++ mpi-linux-x86_64 smp mpicxx --with-production -j6
+        MPICC=mpicc MPICXX=mpicxx \
+             ./build LIBS mpi-linux-x86_64 smp mpicxx --with-production -j6
```
I'm getting a few build errors that seem to get ignored, is that expected? Example:

```
gfortran: error: unrecognized command line option ‘-auto’
Fatal Error by charmc in directory /scratch1/06739/nfischer/spectre/spectre_deps/charm/mpi-linux-x86_64-smp-mpicxx/tmp/libs/ck-libs/ampi
   Command mpif90 -auto -fPIC -I/scratch1/06739/nfischer/spectre/spectre_deps/charm/mpi-linux-x86_64-smp-mpicxx/bin/../include -O2 -c ampifimpl.f90 -o /scratch1/06739/nfischer/spectre/spectre_deps/charm/mpi-linux-x86_64-smp-mpicxx/lib/ampifimpl.o returned error code 1
charmc exiting...
gmake[3]: [/scratch1/06739/nfischer/spectre/spectre_deps/charm/mpi-linux-x86_64-smp-mpicxx/lib/libmoduleampif.a] Error 1 (ignored)
```

Another example:

```
checking dynamic linker characteristics... g++: error: moduleinit275155.o: No such file or directory
ar: creating ../../../../lib/libmodulepose.a
/bin/cp libmodulepose.dep ../../../../lib/.
gmake[3]: Leaving directory `/scratch1/06739/nfischer/spectre/spectre_deps/charm/mpi-linux-x86_64-smp-mpicxx/tmp/libs/ck-libs/pose'
```

wthrowe replied:
Hard to be sure with charm.  The ampifimpl.o build is marked ok-to-fail in the makefile, and I know there are other places where they try to build multiple things and just use whichever works, so might be intentional.  The "-auto" part seems odd in that case, though.  Might be assuming the mpif* compilers are Intel.

The second error is harder to pin down.

Neither seems to cause us problems, in any case.

wthrowe committed:
Move frontera module setup to a script

Changes to the code:
- switch to bash
- set -e
- log the build output
- remove some environment cleanup at the end because the script
  doesn't change the parent environment

wthrowe committed:
Switch to UCX build of charm on frontera

MPI build was giving errors, UCX seems preferred by upstream.

kidder commented:
I can compile successfully, but several tests fail because module h5py cannot be found.  But I don't think that blocks running code...

wthrowe commented:
Please hold off on merging this.  I'm not convinced charm is building correctly, possibly related to the errors @nilsleiffischer pointed out above.

wthrowe committed:
Run tests through charmrun on frontera

wthrowe committed:
Add charm stuff to PATH in frontera

Executables need to be run through charmrun in UCX setups, so it is
nice to have easy access to it.

wthrowe commented:
Significant changes, notably switching to a UCX build of charm.  The remaining test failures are:

* `*.execute_check_output` - missing h5py
* ` Unit.ErrorHandling.FloatingPointExceptions.*` - these actually work, but there's a delay of roughly a minute between the error and the executable exiting, so they time out.
* `"Unit.Visualization.Python.Render1D"` - missing h5py

It would be good to get h5py installed, but I don't know much about python installation so I haven't tried to do that.

nilsvu commented on support/Environments/setup/frontera_gcc.sh:
```diff
@@ -0,0 +1,200 @@
+prepend-path LD_LIBRARY_PATH "$dep_dir/charm/${charm_config}/lib"
+prepend-path CPLUS_INCLUDE_PATH "$dep_dir/charm/${charm_config}/include"
+prepend-path CMAKE_PREFIX_PATH "$dep_dir/charm/${charm_config}/"
+setenv CHARM_VERSION 6.8.2
```
this version is outdated

nilsvu commented on support/Environments/setup/frontera_gcc.sh:
```diff
@@ -0,0 +1,200 @@
+    echo "Installed Charm++ into $dep_dir/charm"
+    cat >$dep_dir/modules/charm <<EOF
+#%Module1.0
+prepend-path PATH "$dep_dir/charm/${charm_config}/bin"
```
I'm getting warnings like this when running executables with `charmrun`:

```
[1641995747.501073] [login1:86419:0]        rc_iface.c:506  UCX  WARN  using maximal value for RETRY_COUNT (7) instead of 40
# (one per PE)
```

Any idea what to do about it?

wthrowe replied:
I see them too.  I think it's a bug in frontera's impi module, which tries to set some UCX RETRY_COUNT settings to 40.  Unless we see communication errors I don't think it's a problem.

nilsvu commented on support/Environments/setup/frontera_gcc.sh:
```diff
@@ -0,0 +1,200 @@
+EOF
+fi
+cd $dep_dir
+
```
Re your question on `h5py`: I suggest you install Py packages in the `$dep_dir` like this:

```sh
# in spectre_setup_modules:
python3 -m venv --prompt spectre --system-site-packages $dep_dir/py_env
source $dep_dir/py_env/bin/activate
# Install h5py with the system HDF5
HDF5_DIR=$TACC_HDF5_DIR pip install --no-binary=h5py h5py
# Install additional Python packages. User can install more using `pip` if they
# want.
pip install yapf==0.29.0
```

Instead of sourcing the `activate` script you can add things to the `PATH` manually if you prefer, e.g. in a module. The `activate` script does only very little. Source the `activate` script or load the module also in `spectre_load_modules` so it gets picked up by cmake.

wthrowe replied:
Thanks!  I've added a module.

nilsvu reviewed (changes requested):
Build works for me, modulo these comments

nilsvu commented on support/Environments/frontera_gcc.sh:
```diff
@@ -252,12 +70,15 @@ spectre_run_cmake() {
     fi
     spectre_load_modules
+    # -D USE_LD=ld - ld.gold seems to hang linking the main executables
     cmake -D CHARM_ROOT=$CHARM_ROOT \
```
I had to add `DEBUG_SYMBOLS=OFF` to fit the build of executables into memory on the head node (which I probably shouldn't have done). Maybe add a message to the user to compile on a compute node, along with the command to allocate one? And/or turn `DEBUG_SYMBOLS` off?

wthrowe replied:
I've added a few sentences about this to the installation instructions, and also mentioned the strange recompilation behavior.

wthrowe committed:
Fix incorrect CHARM_VERSION export

wthrowe committed:
Add h5py and yapf on frontera

wthrowe committed:
Add some frontera installation notes

nilsvu commented on support/Environments/setup/frontera_gcc.sh:
```diff
@@ -0,0 +1,212 @@
+prepend-path PATH ${VIRTUAL_ENV}/bin
+EOF
+HDF5_DIR=$TACC_HDF5_DIR pip install --no-binary=h5py h5py
+pip install yapf==0.29.0
```
(optional) if you add `'pybind11>=2.6.0'` here (note quotes for the `>`) you can set `BUILD_PYTHON_BINDINGS=ON`. Also need to compile `yaml-cpp` with `CMAKE_POSITION_INDEPENDENT_CODE=ON` or with `BUILD_SHARED_LIBS=ON` apparently.

nilsvu reviewed (commented):
LGTM please squash.

wthrowe commented:
Squash what?  There are no fixup commits.

Let's put optional stuff off.  There are already people trying to use frontera and needing this.
